SEP-835: Update authorization spec with default scopes definition (#835)
modelcontextprotocol/modelcontextprotocol: docs/specification/draft/basic/authorization.mdx

MCP servers **MUST** use the HTTP header `WWW-Authenticate` when returning a _401 Unauthorized_ to indicate the location of the resource server metadata URL
as described in [RFC9728 Section 5.1 "WWW-Authenticate Response"](https://datatracker.ietf.org/doc/html/rfc9728#name-www-authenticate-response).

> MCP servers **SHOULD** include a `scope` parameter in the `WWW-Authenticate` header as defined in [RFC 6750 Section 3](https://datatracker.ietf.org/doc/html/rfc6750#section-3) to indicate the scopes required for accessing the resource. This provides clients with immediate guidance on the appropriate scopes to request during authorization.

Author: We're a bit more prescriptive here to make sure that there is consistency :)

Example 401 response with scope guidance:

```http
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Bearer resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource",
                         scope="files:read"
```

MCP clients **MUST** be able to parse `WWW-Authenticate` headers and respond appropriately to `HTTP 401 Unauthorized` responses from the MCP server.

#### Default Scopes

MCP servers **SHOULD** include a `scope` parameter in the `WWW-Authenticate` header as defined in
[RFC 6750 Section 3](https://datatracker.ietf.org/doc/html/rfc6750#section-3) when returning a 401 Unauthorized response.
This provides clients with immediate guidance on the appropriate scopes to request during authorization,
following the principle of least privilege and preventing clients from requesting excessive permissions.

Example 401 response with scope guidance:

```http
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Bearer resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource",
                         scope="files:read"
```

> If no `scope` parameter is provided in the WWW-Authenticate header, clients should fall back to requesting all scopes listed in `scopes_supported` from the Protected Resource Metadata document.

Reviewer: Does the `scope` parameter, if present, have to be a subset of `scopes_supported`? (Seems like yes, but perhaps it should be made explicit.)

Reviewer 2: Because PRM allows resources not to advertise all scopes supported, it seems possible that implementations could expect to differ here. In practice, there isn't any security posture benefit from doing so; the challenge is just as accessible to an unauthenticated client. But I could imagine issues keeping things in sync between the metadata and the authorization logic. Validation would help force a sync, but I'm not sure how necessary that really ends up being.

Author: @goosetav as @mattchenderson called out, this is not necessarily a requirement.

MCP clients **MUST** be able to parse `WWW-Authenticate` headers and respond appropriately to `HTTP 401 Unauthorized` responses from the MCP server.

#### Default Scopes

In addition to the standard Protected Resource Metadata fields defined in [RFC9728](https://datatracker.ietf.org/doc/html/rfc9728),

> MCP servers **MAY** include the optional `scopes_default`extension.

Reviewer 3: ```suggestion
MCP servers **MAY** include the optional `scopes_default` extension.

Reviewer 4: additional nit: The word `extension` here initially made me think this was something bigger (like a separate RFC). Perhaps this would be clearer:

```suggestion
MCP servers **MAY** include an optional `scopes_default` field.
```

> MCP servers **MAY** include the optional `scopes_default` extension.

Reviewer 5: I believe according to [RFC 6750](https://datatracker.ietf.org/doc/html/rfc6750#section-3) it is allowed to return a `scope` field in a `WWW-Authenticate` header for `401` errors as well as `403` errors. 

If the server returned a `scope` parameter in the initial `WWW-Authenticate` at the same time it returned the `resource_metadata` parameter, then the spec wouldn't need to define the `scopes_default` extension and scope handling for both the initial scope selection and the upgrade flow could be united.

Author: That's a really good point, @max-stytch - and it would avoid extra hassle with extending the PRM. Updated the doc.

`scopes_default` is a JSON array containing a list of scope values that represent the default scopes for clients to use
when initiating the authorization flow. This field provides guidance to clients about the appropriate scopes to request
when starting the authorization process or responding to scope-related errors.

MCP servers **SHOULD** include this field to follow the principle of least privilege and prevent clients from requesting excessive permissions.
By providing `scopes_default`, servers guide clients toward requesting only the minimum scopes necessary for typical operations,
rather than all available scopes.

> The `scopes_default` values **MUST** be a subset of the `scopes_supported` values.

Reviewer 6: In a typical implementation, how to determine the `scopes_default`?

Once the client discovers the available tools, tool selection and execution will be dynamic based on the reasoning at the host side.

Author: Resolving since the suggested approach changed. Recommend checking #850 for exploration on primitive authorization.

MCP servers **MUST** use the HTTP header `WWW-Authenticate` when returning a _401 Unauthorized_ to indicate the location of the resource server metadata URL
as described in [RFC9728 Section 5.1 "WWW-Authenticate Response"](https://datatracker.ietf.org/doc/html/rfc9728#name-www-authenticate-response).

MCP servers **SHOULD** include a `scope` parameter in the `WWW-Authenticate` header as defined in
[RFC 6750 Section 3](https://datatracker.ietf.org/doc/html/rfc6750#section-3)
to indicate the scopes required for accessing the resource. This provides clients with immediate
guidance on the appropriate scopes to request during authorization,
following the principle of least privilege and preventing clients from requesting excessive permissions.

Example 401 response with scope guidance:

```http
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Bearer resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource",
                         scope="files:read"
```

MCP clients **MUST** be able to parse `WWW-Authenticate` headers and respond appropriately to `HTTP 401 Unauthorized` responses from the MCP server.

> If no `scope` parameter is provided in the `WWW-Authenticate` header, clients should fall back to requesting all

Reviewer 7: ```suggestion
If no `scope` parameter is provided in the `WWW-Authenticate` header, clients SHOULD fall back to requesting all
```

> If no `scope` parameter is provided in the `WWW-Authenticate` header, clients SHOULD fall back to requesting all

Reviewer 8: This is a duplicate of the "scope selection strategy" section; perhaps this could just be a pointer to that section.

   - At the path of the server's MCP endpoint: `https://example.com/public/mcp` could host metadata at `https://example.com/.well-known/oauth-protected-resource/public/mcp`
   - At the root: `https://example.com/.well-known/oauth-protected-resource`

MCP clients **MUST** support both discovery mechanisms and use the resource metadata URL from the parsed `WWW-Authenticate` headers when present; otherwise, they **MUST** fall back to constructing and requesting the well-known URIs in the order listed above.

MCP servers **SHOULD** include a `scope` parameter in the `WWW-Authenticate` header as defined in
[RFC 6750 Section 3](https://datatracker.ietf.org/doc/html/rfc6750#section-3)
to indicate the scopes required for accessing the resource. This provides clients with immediate
guidance on the appropriate scopes to request during authorization,
following the principle of least privilege and preventing clients from requesting excessive permissions.

> The scopes included in the `WWW-Authenticate` challenge **MAY** match `scopes_supported`, be a subset

Reviewer 9: This is interesting. I am curious why this isn't a SHOULD. It feels to me that the scopes asked for should always be a subset of scopes_supported.

   OAuth client themselves (e.g., through a configuration interface hosted by the
   server).

#### Scope Selection Strategy

When implementing authorization flows, MCP clients **SHOULD** follow the principle of least privilege by requesting
only the scopes necessary for their intended operations. During the initial authorization handshake, MCP clients
**SHOULD** follow this priority order for scope selection:

1. **Use `scopes_default`** from the Protected Resource Metadata document as the starting point

> 2. **If `scopes_default` is not available**, use all scopes defined `scopes_supported`

Reviewer 4: ```suggestion
2. **If `scopes_default` is not available**, use all scopes defined in `scopes_supported`
```

> 2. **If `scopes_default` is not available**, use all scopes defined in `scopes_supported`

Reviewer 6: Wouldn't requesting/issuing a token with all the supported scopes result in an overly scoped token that increases the security risk, in case the token is compromised?

Author: @ayshsandu - if the server doesn't provide any other clues as to what scopes should be requested, the client alone can't determine the "minimum required" and has to use what's supported.

> Clients **MAY** make individual decisions on scope selection and rely on incremental consent.

Reviewer 7: ```suggestion
Clients **MAY** use out-of-band information to decide on scope selection and incrementally ask the user for consent.
```

   OAuth client themselves (e.g., through a configuration interface hosted by the
   server).

#### Scope Selection Strategy

When implementing authorization flows, MCP clients **SHOULD** follow the principle of least privilege by requesting
only the scopes necessary for their intended operations. During the initial authorization handshake, MCP clients
**SHOULD** follow this priority order for scope selection:

1. **Use `scope` parameter** from the initial `WWW-Authenticate` header in the 401 response, if provided

> 2. **If `scope` is not available**, use scopes defined in `scopes_supported` from the Protected Resource Metadata document

Reviewer 7: ```suggestion
2. **If `scope` is not available**, use all scopes defined in `scopes_supported` from the Protected Resource Metadata document
```

> When no `scope` parameter is provided in the `WWW-Authenticate` header, clients **SHOULD** request all scopes from `scopes_supported`. This approach accommodates the general-purpose nature of MCP clients, which typically lack domain-specific knowledge to make informed decisions about individual scope selection. Requesting all available scopes allows the authorization server and end-user to determine appropriate permissions during the consent process.

Reviewer 8: ```suggestion
This approach accommodates the general-purpose nature of MCP clients, which typically lack domain-specific knowledge to make informed decisions about individual scope selection. Requesting all available scopes allows the authorization server and end-user to determine appropriate permissions during the consent process.
```
First sentence duplicates the previous list, or you could possibly remove the whole paragraph, it's kind of reiterate two paragraphs down.

> Clients **MAY** use out-of-band information to decide on scope selection and incrementally ask the user for consent.

Reviewer 5: > in-band method for mapping scopes to tools at some point 

I don't think that in-band scope/tool mapping at `tools/list` time is a tractable problem with the rise of "everything" tools like [Square's](https://github.com/square/square-mcp-server?tab=readme-ov-file#tool-reference) and [Stainless's](https://www.stainless.com/docs/guides/generate-mcp-server-from-openapi#enabling-dynamic-tools-for-larger-apis) - oftentimes the permissions needed are a function of the tool call arguments instead of the tool name itself

> Clients **MAY** use out-of-band information to decide on scope selection and incrementally ask the user for consent.

Reviewer 10: The whole proposal seems to assume that coarse-grained scopes are worth supporting. It doesn't mean FGA cannot also be supported (now or later).

To be clear, I'm in favor of `WWW-Authenticate` header support, I'm just speaking up for more. I don't think Square's implementation should be used as an argument for not proactively sharing authorization requirements. It's possible to do with that scheme, just not within the current confines of the protocol. The thought exercise is: how could we better support that scheme within the protocol so it's standardized?

> For any scenario that requires fine-grained authorization, a scopes-to-tools mechanism would become outdated as well.

It would not apply in an FGA scenario which is different from being outdated. The entire concept of authorization including coarse-grained is not a MUST. I think coarse-grained scopes will be around for a long time.

These things are at odds with one another:
> preventing clients from requesting excessive permissions

vs

> **Extended approach:** Include existing scopes, newly required scopes, and related scopes that commonly work together
> **User-friendly approach:** Group commonly used related scopes together to reduce subsequent authorization flows

The server is forced to guess at what's appropriate when making a suggestion. A goldilocks problem with tension between security vs a good UX. With a little info up front, the client could predict what it requires more accurately.
https://github.com/modelcontextprotocol/modelcontextprotocol/issues/1385
@localden, did a [debate on your proposal here](https://docs.google.com/document/d/1460o7LRZPMDFxoDgdYTI4gdxjgdBMKDxnN5Ic1DGhng/edit?tab=t.0) play out anywhere?

> Clients **MAY** use out-of-band information to decide on scope selection and incrementally ask the user for consent.

Author: @jbaldo yeah, whether we like it or not, scopes are already part of the spec and are a standard way of managing client access, so we have to support it. There is more to this as an end-to-end story, of course, but a de-facto scope-based implementation can instantly unlock a lot of auth-related scenarios that previously relied on "lemme request everything from the get go" approaches.

   OAuth client themselves (e.g., through a configuration interface hosted by the
   server).

#### Scope Selection Strategy

When implementing authorization flows, MCP clients **SHOULD** follow the principle of least privilege by requesting
only the scopes necessary for their intended operations. During the initial authorization handshake, MCP clients
**SHOULD** follow this priority order for scope selection:

1. **Use `scope` parameter** from the initial `WWW-Authenticate` header in the 401 response, if provided
2. **If `scope` is not available**, use scopes defined in `scopes_supported` from the Protected Resource Metadata document

> Clients **MAY** use one, several, or all scopes from `scopes_supported` if `scope` is not available.

Reviewer 11: Given the fact that MCP Clients are designed to be general purpose and usable with any MCP Server, this means they won't inherently understand the scopes and whether to choose one, several, or all. Can we give more prescriptive guidance?

I recall the original debate on this where I was curious why the MCP Client needed to send scopes at all, given it's lack of understanding of their meaning. The reasoning was that many authorization servers require scopes today and therefore we can encourage faster MCP adoption by meeting the patterns of existing AS's and sending the full list of known scopes. The end-user can select which to grant during consent. 

If that remains the case, then perhaps we learn into recommending clients send "all scopes"?

Reviewer 11: A simple fix could be an additional sentence: 
_"Implementations that lack domain-specific knowledge to inform scope selection SHOULD request all available scopes from `scopes_supported`."_

Or a non-normative variation:
_"Implementation note: When no other information is available to inform scope selection, requesting all scopes from `scopes_supported` is an acceptable approach."_

   OAuth client themselves (e.g., through a configuration interface hosted by the
   server).

#### Scope Selection Strategy

When implementing authorization flows, MCP clients **SHOULD** follow the principle of least privilege by requesting
only the scopes necessary for their intended operations. During the initial authorization handshake, MCP clients
**SHOULD** follow this priority order for scope selection:

1. **Use `scope` parameter** from the initial `WWW-Authenticate` header in the 401 response, if provided
2. **If `scope` is not available**, use all scopes defined in `scopes_supported` from the Protected Resource Metadata document, omitting the `scope` parameter if `scopes_supported` is undefined.

This approach accommodates the general-purpose nature of MCP clients, which typically lack domain-specific knowledge to make informed decisions about individual scope selection. Requesting all available scopes allows the authorization server and end-user to determine appropriate permissions during the consent process.

> Clients **MAY** use out-of-band information to decide on scope selection and incrementally ask the user for consent.

Reviewer 9: Do we need to write this down? I think everything not specified is implicitly allowed

> Clients **MAY** use out-of-band information to decide on scope selection and incrementally ask the user for consent.

Reviewer 9: I am not sure why we would want to include this. The absence of it means that clients can do whatever they want anyway. I am not sure we need to spell it out, but i am okay iwth leaving it .

| 403         | Forbidden    | Invalid scopes or insufficient permissions |
| 400         | Bad Request  | Malformed authorization request            |

#### Scope Error Handling

> This section covers handling insufficient scope errors during runtime operations when a client already has a token but needs additional permissions. This follows the error handling patterns defined in [RFC 6750 (Bearer Token Usage)](https://datatracker.ietf.org/doc/html/rfc6750) and leverages the metadata fields from [RFC 9728 (OAuth 2.0 Protected Resource Metadata)](https://datatracker.ietf.org/doc/html/rfc9728).

Reviewer 7: Not sure it matters much, but the Bearer Token Usage spec is getting rolled into [Section 5 of OAuth 2.1](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-13#section-5) in case you wanted to use a more up to date reference.

Reviewer 3: It's a good idea. I'd suggest opening a tracking issue to update when the I-D is ratified.

##### Runtime Insufficient Scope Errors

When a client makes a request with insufficient authorization scope during runtime operations, the server **SHOULD** respond with:

- `HTTP 403 Forbidden` status code (per [RFC 6750 Section 3.1](https://datatracker.ietf.org/doc/html/rfc6750#section-3.1))
- `WWW-Authenticate` header with the `Bearer` scheme and additional parameters:
  - `error="insufficient_scope"` - indicating the specific type of authorization failure
  - `scope="required_scope1 required_scope2"` - specifying the minimum scopes needed for the operation
  - `error_description` (optional) - human-readable description of the error

Example insufficient scope response:

```http
HTTP/1.1 403 Forbidden
WWW-Authenticate: Bearer error="insufficient_scope",
                         scope="files:read files:write",
                         error_description="Additional file write permission required"
```

##### Client Scope Upgrade Flow

When a client receives any scope-related error (initial authorization or runtime insufficient scope),
it **SHOULD** follow this upgrade flow:

1. **Parse error information** from the authorization server response or `WWW-Authenticate` header

> 2. **Determine required scopes** using the following fallback hierarchy:

Reviewer 7: ```suggestion
2. **Determine required scopes** using the following algorithm:
```

>    - If the `scope` parameter is present in the `WWW-Authenticate` header, use those scopes

Reviewer 7: Should it request _only_ those scopes? or add those scopes to the previously requested scopes? Since there isn't a way in OAuth to tie an authorization request to a previous one, the AS might issue a token with _only_ these requested scopes that might then fail on other operations. The client would be in a position to need to maintain two access tokens and the scopes associated with them in order to continue normal operation.

Reviewer 3: I'd rephrase it this way:
```suggestion
If the `scope` parameter is present in the `WWW-Authenticate` header, include those scopes

Reviewer 8: I think it'd be interesting to consider how these flows might work within clients, and how that affects this proposal.  A naive approach in say, the Claude Desktop app would just keep my access token forever with upgraded scopes (assuming it gets refresh tokens with upgraded scopes as well).  This probably doesn't fit well with the motivations for adding the ability to have scope upgrades.

It's not necessarily even about the token being compromised, the user (or server) may just prefer that the upgraded scopes are renewed daily (for example).  Perhaps this could be addressed with my suggestion below about (not) validating scope upgrades?

   - If `scope` is not available, consult the `scopes_default` field from the Protected Resource Metadata document
   - If `scopes_default` is not available, use all scopes listed in `scopes_supported` from the Protected Resource Metadata document
3. **Validate scope upgrade** to prevent infinite loops:
   - Calculate the union of current scopes and newly required scopes
   - If the resulting scope set is identical to the previously requested scopes, the client **MUST NOT** retry the authorization
   - If no new scopes can be added (all required scopes were already requested), the client **SHOULD** treat this as a permanent authorization failure
4. **Initiate (re-)authorization** with the broader scope set only if new scopes were identified

> 5. **Retry the original request** with the new authorization

Reviewer 3: I'd like clients not DDoS IdPs.
```suggestion
5. **Retry the original request** with the new authorization no more than a few times and treat this as a permanent authorization failure

| 403         | Forbidden    | Invalid scopes or insufficient permissions |
| 400         | Bad Request  | Malformed authorization request            |

#### Scope Error Handling

This section covers handling insufficient scope errors during runtime operations when a client already has a token but needs additional permissions. This follows the error handling patterns defined in [OAuth 2.1 Section 5](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-13#section-5) and leverages the metadata fields from [RFC 9728 (OAuth 2.0 Protected Resource Metadata)](https://datatracker.ietf.org/doc/html/rfc9728).

##### Runtime Insufficient Scope Errors

When a client makes a request with an access token with insufficient scope during runtime operations, the server **SHOULD** respond with:

- `HTTP 403 Forbidden` status code (per [RFC 6750 Section 3.1](https://datatracker.ietf.org/doc/html/rfc6750#section-3.1))

> - `WWW-Authenticate` header with the `Bearer` scheme and additional parameters:

Reviewer 8: For consistency with the 401 response, should this include the resource metadata URI as well?  Otherwise clients need to keep track of that somewhere.

  - `error="insufficient_scope"` - indicating the specific type of authorization failure
  - `scope="required_scope1 required_scope2"` - specifying the minimum scopes needed for the operation
  - `error_description` (optional) - human-readable description of the error

Example insufficient scope response:

```http
HTTP/1.1 403 Forbidden
WWW-Authenticate: Bearer error="insufficient_scope",
                         scope="files:read files:write",
                         error_description="Additional file write permission required"
```

##### Client Scope Upgrade Flow

When a client receives any scope-related error (initial authorization or runtime insufficient scope),
it **SHOULD** follow this upgrade flow:

1. **Parse error information** from the authorization server response or `WWW-Authenticate` header
2. **Determine required scopes** using the following algorithm:
   - If the `scope` parameter is present in the `WWW-Authenticate` header, use those scopes
   - If `scope` is not available, use all scopes listed in `scopes_supported` from the Protected Resource Metadata document

> 3. **Validate scope upgrade** to prevent infinite loops:

Reviewer 8: This rules out a possibility for timed access (or some other condition), where the scope is no longer available for a given access token, but reauthorizing would re-establish the scope.  It feels a bit like "second guessing" the server a bit, and the "limited retry" logic below would be enough to detect a truly broken setup.

- **Minimum approach**: Include only the newly required scopes for the specific operation
- **Recommended approach**: Include both existing relevant scopes and newly required scopes to prevent clients from losing previously granted permissions
- **Extended approach**: Include existing scopes, newly required scopes, and related scopes that commonly work together

The choice depends on the server's assessment of user experience impact and authorization friction. Servers **SHOULD** be consistent in their scope inclusion strategy to provide predictable behavior for clients.

Servers **SHOULD** consider the user experience impact when determining which scopes to include in the response, as misconfigured scopes may require frequent user interaction.

Example insufficient scope response:

```http
HTTP/1.1 403 Forbidden
WWW-Authenticate: Bearer error="insufficient_scope",
                         scope="files:read files:write user:profile",
                         resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource",
                         error_description="Additional file write permission required"
```

> ##### Client Scope Upgrade Flow

Reviewer 12: Wouldn't a better name for this be **Progressive Scoping Flow**?

Author: I see this is what you folks are [using at Descope](https://www.descope.com/blog/post/progressive-scoping), but I am not sure if this is an industry term - in past experiences on auth libraries, this was called "scope upgrade" - I am actually somewhat impartial to the name here. @aaronpk @D-McAdams - any strong feelings either way?

Reviewer 5: Dick Hardt has pointed out this has been called [incremental authorization](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/284#issuecomment-2788291650) in the past

Clients will receive scope-related errors during initial authorization or at runtime (`insufficient_scope`).
Clients may respond to these errors by requesting a new access token with an increased set of scopes via an upgrade flow.
Clients acting on behalf of a user **SHOULD** attempt the upgrade flow. Clients acting on their own behalf (`client_credentials` clients)
**MAY** attempt the upgrade flow or abort the request immediately.

The upgrade flow is as follows:

1. **Parse error information** from the authorization server response or `WWW-Authenticate` header
2. **Determine required scopes** using the following algorithm:
   - If the `scope` parameter is present in the `WWW-Authenticate` header, use those scopes exactly as specified (trusting the server to include any necessary existing scopes along with newly required scopes)
   - If `scope` is not available, use all scopes listed in `scopes_supported` from the Protected Resource Metadata document
3. **Initiate (re-)authorization** with the determined scope set

> 4. **Retry the original request** with the new authorization no more than a few times and treat this as a permanent authorization failure

Reviewer 8: Without trying to open up a can of worms -- how should "a permanent authorization failure" be treated: should the access token be thrown away entirely, or should the current call just fail and keep the token to try again with a different request later?

I also wonder if this is really any different to any other failed request -- how are retries handled elsewhere in the spec?  The caching paragraph under this is only really necessary to implement this retry logic, and I guess how that's implemented could be left open.  For a very simple client it may be quite straightforward to notice it's failed twice, without a cache.

Author: Simplified that entire section @simonrussell - thank you for the feedback!

Clients **SHOULD** implement retry limits and **SHOULD** cache scope upgrade attempts to avoid
repeated failures for the same resource and operation combination. The cache should track failed
authorization attempts by resource and the scopes that were denied, preventing clients from
repeatedly requesting the same permissions that were previously denied and avoiding unnecessary

> authorization server requests for known-failing scenarios.

Reviewer 13: I was thinking about this cache today. I'm wondering what the granularity should be...

So we have:
* HTTP method
* MCP method (call/tool)
* Specific tool being called
* Specific payload of specific tool being called

Really any of these could be the reason for a new set of scopes... how could a client really keep track of this?

**MAY** attempt the upgrade flow or abort the request immediately.

The upgrade flow is as follows:

1. **Parse error information** from the authorization server response or `WWW-Authenticate` header
2. **Determine required scopes** using the following algorithm:
   - If the `scope` parameter is present in the `WWW-Authenticate` header, use those scopes
   - If `scope` is not available, use all scopes listed in `scopes_supported` from the Protected Resource Metadata document
3. **Calculate the union** of current scopes and newly required scopes
4. **Initiate (re-)authorization** with the scope set
5. **Retry the original request** with the new authorization no more than a few times and treat this as a permanent authorization failure

Clients **SHOULD** implement retry limits and **SHOULD** cache scope upgrade attempts to avoid repeated failures
for the same resource and operation combination. The cache should track failed authorization attempts by resource
and the scopes that were denied, preventing clients from repeatedly requesting the same permissions that were

> previously denied and avoiding unnecessary authorization server requests for known-failing scenarios.

Reviewer 13: What is the new default for tool calls after you hit one that needed more auth? Is it the union scopes or the minimum scopes?

and leverages the metadata fields from [RFC 9728 (OAuth 2.0 Protected Resource Metadata)](https://datatracker.ietf.org/doc/html/rfc9728).

##### Runtime Insufficient Scope Errors

When a client makes a request with an access token with insufficient
scope during runtime operations, the server **SHOULD** respond with:

- `HTTP 403 Forbidden` status code (per [RFC 6750 Section 3.1](https://datatracker.ietf.org/doc/html/rfc6750#section-3.1))
- `WWW-Authenticate` header with the `Bearer` scheme and additional parameters:
  - `error="insufficient_scope"` - indicating the specific type of authorization failure
  - `scope="required_scope1 required_scope2"` - specifying the minimum scopes needed for the operation
  - `resource_metadata` - the URI of the Protected Resource Metadata document (for consistency with 401 responses)
  - `error_description` (optional) - human-readable description of the error

**Server Scope Management**: When responding with insufficient scope errors, servers
**SHOULD** include the scopes needed to satisfy the current request in the `scope`
parameter. Servers have flexibility in determining which scopes to include:

> - **Minimum approach**: Include only the newly required scopes for the specific operation

Reviewer 14: I think this is slightly inconsistent with this line below:
> (trusting the server to include any necessary existing scopes along with newly required scopes)

Maybe we just delete this line?

Reviewer 4: @pcarleton Yes, I would love to provide an opinionated minimum here. In my experience there is a _lot_ of variability in how Authorization Servers treat existing scopes - the AS will know what its _own_ policy around existing scopes will be, but we obviously can't expect the client to know that policy.

Perhaps:
```suggestion
- **Minimum approach**: Include the newly-required scopes for the specific operation. Include any existing granted scopes as well, if they are required prevent clients from losing previously granted permissions.
```

| 403         | Forbidden    | Invalid scopes or insufficient permissions |
| 400         | Bad Request  | Malformed authorization request            |

#### Scope Error Handling

> This section covers handling insufficient scope errors during runtime operations when a client already has a token but needs additional permissions. This follows the error handling patterns defined in [OAuth 2.1 Section 5](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-13#section-5) and leverages the metadata fields from [RFC 9728 (OAuth 2.0 Protected Resource Metadata)](https://datatracker.ietf.org/doc/html/rfc9728).

Reviewer 15: @ciamshrek tell me more. What would be the user flow in this scenario?

##### Runtime Insufficient Scope Errors

When a client makes a request with an access token with insufficient scope during runtime operations, the server **SHOULD** respond with:

- `HTTP 403 Forbidden` status code (per [RFC 6750 Section 3.1](https://datatracker.ietf.org/doc/html/rfc6750#section-3.1))
- `WWW-Authenticate` header with the `Bearer` scheme and additional parameters:
  - `error="insufficient_scope"` - indicating the specific type of authorization failure
  - `scope="required_scope1 required_scope2"` - specifying the minimum scopes needed for the operation
  - `error_description` (optional) - human-readable description of the error

Example insufficient scope response:

```http
HTTP/1.1 403 Forbidden
WWW-Authenticate: Bearer error="insufficient_scope",
                         scope="files:read files:write",
                         error_description="Additional file write permission required"
```

##### Client Scope Upgrade Flow

When a client receives any scope-related error (initial authorization or runtime insufficient scope),

> it **SHOULD** follow this upgrade flow:

Reviewer 16: I may be wrong, but I think there are at least 3 levels of definitions of the client vs server differentiation.
When the level is not specified, it's possible to misunderstand the specification by thinking about a different level:

1. HTTP client and server - here the client is the side that initiates the connection, so an HTTP server cannot send a request (see https://github.com/modelcontextprotocol/modelcontextprotocol/discussions/1077#discussioncomment-13914502)
2. JSON-RPC client and server - here the client is defined as the origin of request objects and the handler of response objects, so a JSON-RPC server cannot send a request (see https://www.jsonrpc.org/specification)
3. MCP client and server - here the roles are not tied to the HTTP or JSON-RPC protocols, but defined in terms of capabilities, so the client offers "sampling, roots, elicitation" and server offers "resources, prompts, tools" (see https://modelcontextprotocol.io/specification/2025-06-18#features)

> it **SHOULD** follow this upgrade flow:

Author: @kentcdodds yes, this is consistent - it effectively would give you a response with the clue around which scopes are needed.

and leverages the metadata fields from [RFC 9728 (OAuth 2.0 Protected Resource Metadata)](https://datatracker.ietf.org/doc/html/rfc9728).

##### Runtime Insufficient Scope Errors

When a client makes a request with an access token with insufficient
scope during runtime operations, the server **SHOULD** respond with:

- `HTTP 403 Forbidden` status code (per [RFC 6750 Section 3.1](https://datatracker.ietf.org/doc/html/rfc6750#section-3.1))
- `WWW-Authenticate` header with the `Bearer` scheme and additional parameters:
  - `error="insufficient_scope"` - indicating the specific type of authorization failure
  - `scope="required_scope1 required_scope2"` - specifying the minimum scopes needed for the operation
  - `resource_metadata` - the URI of the Protected Resource Metadata document (for consistency with 401 responses)
  - `error_description` (optional) - human-readable description of the error

**Server Scope Management**: When responding with insufficient scope errors, servers
**SHOULD** include the scopes needed to satisfy the current request in the `scope`
parameter.

Servers have flexibility in determining which scopes to include:

> - **Minimum approach**: Include the newly-required scopes for the specific operation. Include any existing granted scopes as well, if they are required prevent clients from losing previously granted permissions.

Reviewer 17: This is missing the word "to". I think it may also be missing a comma but I'm not sure which is the intended meaning:

```suggestion
- **Minimum approach**: Include the newly-required scopes for the specific operation. Include any existing granted scopes as well, if they are required, to prevent clients from losing previously granted permissions.
```

```suggestion
- **Minimum approach**: Include the newly-required scopes for the specific operation. Include any existing granted scopes as well, if they are required to prevent clients from losing previously granted permissions.
```

- **Recommended approach**: Include both existing relevant scopes and newly required scopes to prevent clients from losing previously granted permissions
- **Extended approach**: Include existing scopes, newly required scopes, and related scopes that commonly work together

The choice depends on the server's assessment of user experience impact and authorization friction.
Servers **SHOULD** be consistent in their scope inclusion strategy to provide predictable behavior for clients.

Servers **SHOULD** consider the user experience impact when determining which scopes to include in the
response, as misconfigured scopes may require frequent user interaction.

Example insufficient scope response:

```http
HTTP/1.1 403 Forbidden
WWW-Authenticate: Bearer error="insufficient_scope",
                         scope="files:read files:write user:profile",
                         resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource",
                         error_description="Additional file write permission required"
```

##### Step-Up Authorization Flow

Clients will receive scope-related errors during initial authorization or at runtime (`insufficient_scope`).

> Clients may respond to these errors by requesting a new access token with an increased set of scopes via a step-up authorization flow.

Reviewer 9: ```suggestion
Clients **SHOULD** respond to these errors by requesting a new access token with an increased set of scopes via a step-up authorization flow or handle the errors in other, appropriate ways.
```

to explicitly specify the target resource for which the token is being requested. This requirement aligns with the recommendation in
[RFC 9728 Section 7.4](https://datatracker.ietf.org/doc/html/rfc9728#section-7.4). This ensures that access tokens are bound to their intended resources and
cannot be misused across different services.

### Scope Minimization

> Following the principle of least privilege, MCP clients **SHOULD** request only the minimum scopes necessary for their intended operations.

Reviewer 12: When the MCP client will refresh the token with the authorization server, they can also "downscope" and request tokens that are less than what is required at the time. As long as the consent has been granted for a subset of the scopes in the original (expired) access token, the OAuth authorization server should (assuming they support this) return a new token with those reduced scopes. Therefore it's important to mention the following:

Authorization Servers are responsible for evaluating requested scopes against prior consent and MAY issue access tokens with narrower scopes, as permitted by [RFC 6749 §1.5](https://datatracker.ietf.org/doc/html/rfc6749#section-1.5). This enables secure, progressive access, allowing clients to adjust scope requirements over time, while adhering to the principle of least privilege.

> The `scopes_default` extension in Protected Resource Metadata provides server guidance for appropriate default scope selection, helping clients avoid:

Reviewer 4: ```suggestion
The `scopes_default` field in the authorization server's Protected Resource Metadata document provides server guidance for appropriate default scope selection, helping clients avoid:
```

Author: @nbarbettini - but the AS is not the one that provides the PRM, it's the MCP server?

> - **Over-privileged tokens**: Requesting excessive scopes increases the potential impact if tokens are compromised

Author: @tolginator - same as @nbarbettini, I am not sure that is the case here, because if the RP is compromised, you can do so much more damage than you could _with a token_ that is used to access said RP.

- **User friction**: Large scope requests may cause user concern or rejection during authorization

> - **Attack surface**: Tokens with unnecessary privileges expand the potential attack surface

Reviewer 4: This feels like it duplicates the "Over-privileged tokens" bullet to me.

MCP servers **SHOULD** implement fine-grained scopes that allow clients to request access to only the specific resources and operations they need.

> Authorization servers **SHOULD** clearly communicate the implications of each scope to users during the authorization process.

Reviewer 3: I wonder if more prescriptive guidance is warranted lest this is easily ignored. As in, what is a clear communication channel a developer would need to implement?
I don't have a great suggestion other than mumble something about _clear consent dialog_ where most people would click the _OK_ button, and am looking for feedback here.
